includes/acl/sjson/sjson_parser.h

```diff
@@ -0,0 +1,719 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "sjson_parser_error.h"
+
+#include <cctype>
+#include <cmath>
+#include <cstring>
+
+namespace acl
+{
+	class SJSONParser
+	{
+	public:
+		SJSONParser(const char* const input, int input_length)
+			: m_input(input),
+			  m_input_length(input_length)
+		{
+			init_state();
+		}
+
+		bool object_begins()
+		{
+			return read_opening_brace();
+		}
+
+		bool object_begins(const char* const having_name)
+		{
+			return read_key(having_name) && read_equal_sign() && object_begins();
+		}
+
+		bool object_ends()
+		{
+			return read_closing_brace();
+		}
+
+		bool array_begins(const char* const having_name)
+		{
+			return read_key(having_name) && read_equal_sign() && read_opening_bracket();
+		}
+
+		bool try_array_begins(const char* const having_name)
+		{
+			State s = save_state();
+
+			if (!array_begins(having_name))
+			{
+				restore_state(s);
+				return false;
+			}
+
+			return true;
+		}
+
+		bool array_begins()
+		{
+			return read_opening_bracket();
+		}
+
+		bool try_array_ends() 
+		{
+			State s = save_state();
+
+			if (!array_ends())
+			{
+				restore_state(s);
+				return false;
+			}
+
+			return true;
+		}
+
+		bool array_ends()
+		{
+			return read_closing_bracket();
+		}
+
+		bool read(const char* const key, const char*& value, int& length)
+		{
+			return read_key(key) && read_equal_sign() && read_string(value, length);
+		}
+
+		bool try_read(const char* const key, const char*& value, int& length)
+		{
+			State s = save_state();
+			
+			if (!read(key, value, length))
+			{
+				restore_state(s);
+
+				value = nullptr;
+				length = 0;
+
+				return false;
+			}
+
+			return true;
+		}
+
+		bool read(const char* const key, bool& value)
+		{
+			return read_key(key) && read_equal_sign() && read_bool(value);
+		}
+
+		bool read(const char* const key, double& value)
+		{
+			return read_key(key) && read_equal_sign() && read_double(value);
+		}
+
+		bool read(const char* const key, int& value)
+		{
+			return read_key(key) && read_equal_sign() && read_int(value);
+		}
+
+		bool read(const char* const key, double* const values, int num_elements)
+		{
+			return read_key(key) && read_equal_sign() && read_opening_bracket() && read(values, num_elements) && read_closing_bracket();
+		}
+
+		bool try_read(const char* const key, double* const values, int num_elements)
+		{
+			State s = save_state();
+			
+			if (!read(key, values, num_elements))
+			{
+				restore_state(s);
+
+				for (int i = 0; i < num_elements; ++i)
+				{
+					values[i] = 0.0;
+				}
+
+				return false;
+			}
+
+			return true;
+		}
+
+		bool read(double* const values, int num_elements)
```

Comment:
I feel a function like this might be better off private or protected rather than public since we already have the other two variants that validate and call this